### DISCUSSION TITLE
[RFC-1] Generate typealias for Selection types

### DISCUSSION THREAD

**Discussion by maticzav** [maticzav/swift-graphql#5] [Ideas]
It would be very beneficial if instead of having to write `Selection<TypeLock, Target>`, you could write `SType<Target>`. We can generate bindings for all the internal types since those are of finite set magnitude, and reduce the complexity drastically.

To showcase the idea, here's a rewritten example from our docs with the new implementation.

```swift
import SwiftGraphQL

// Define a Swift model.
struct Human: Identifiable {
    let id: String
    let name: String
    let homePlanet: String?
}

// Create a selection.
let human = SGHuman<Human> {
    Human(
        id: try $0.id(), 
        name: try $0.name(),
        homePlanet: try $0.homePlanet()
    )
}

// Construct a query.
let query = SGQuery<[Human]> {
    try $0.humans(human.list)
}

// Perform the query.
SG.send(query, to: "http://swift-graphql.heroku.com") { result in
    if let data = try? result.get() {
        print(data) // [Human]
    }
}
```

Open questions:

- [ ] What naming conventions do we want to follow to prevent naming collisions?

**Comment by Amzd** [comment]
You could make them children of Selection, it would not be that much shorter but you can use the dot syntax which autocompletes rather than the `<` & `>`. Another feature of this is that it can infer the return type from the closure.

Example with typealias on Selection:
```swift
// Create a selection.
let human = Selection.Human {
    Human(
        id: try $0.id(), 
        name: try $0.name(),
        homePlanet: try $0.homePlanet()
    )
}

// Construct a query.
let query = Selection.Query {
    try $0.humans(human.list)
}
```

To make this work we need to generate this:
```swift
extension Selection where TypeLock == Never, Type == Never {
    typealias Human<T> = Selection<T, Objects.Human>
    typealias Query<T> = Selection<T, Operations.Query>
}
```

**Reply by Amzd** [reply]
This would have to make sure that the Objects type is never a duplicate of an Operations type.

**Reply by maticzav** [reply]
I don’t understand the comment.

**Reply by Amzd** [reply]
If we generate the bottom part then the example in the top part works. I edited to make it more clear. Let me know what I didn't make clear enough

**Reply by Amzd** [reply]
This is what I meant with autocomplete btw 
<img width="369" alt="image" src="https://user-images.githubusercontent.com/17001151/106723844-d3d9c280-6607-11eb-8f15-1b9b706b06fb.png">

**Reply by maticzav** [reply]
Yes, wonderful! I get it now. Will include it in the beta release.

Could you explain the benefit of having `Type == Never` at the top? Is there any difference if we just omit that part altogether?

**Reply by Amzd** [reply]
Not sure if it is related but in my test project without Never the autocompletion fails: 
| Without Never | Never |
| -------- | -------- |
| <img width="576" alt="image" src="https://user-images.githubusercontent.com/17001151/106755500-7efd7280-662e-11eb-9065-ad58ebabd940.png"> | <img width="562" alt="image" src="https://user-images.githubusercontent.com/17001151/106755564-9177ac00-662e-11eb-9c63-f69bce881b08.png"> |

https://user-images.githubusercontent.com/17001151/106756234-5c1f8e00-662f-11eb-83fb-f53c7c4d4638.mov



Which would make sense because it is unable to figure out the type of Selection

**Reply by Amzd** [reply]
To Add: it's basically saying the first Selection doesn't need it's Type and TypeLock. we specify it so the user doesn't have to type:

```swift
Selection<Never, Never>.Query
```

**Reply by maticzav** [reply]
This is great! Thanks!
